File: agent/tools/figma_mcp.py
"""Figma MCP tools for accessing Figma design files.

This module provides tools to interact with Figma design files via the
Framelink MCP server (https://github.com/GLips/Figma-Context-MCP).
Uses the `figma-developer-mcp` npm package launched via stdio transport,
so no persistent HTTP connection is required.

Set FIGMA_API_KEY to a Figma Personal Access Token to enable the tools.
"""

import os
from typing import Any

from langchain_mcp_adapters.client import MultiServerMCPClient


def _unwrap_exception(e: Exception) -> str:
    """Extract a readable message from a plain or grouped exception."""
    if isinstance(e, BaseExceptionGroup):
        return "; ".join(str(sub) for sub in e.exceptions)
    return str(e)


def _make_client() -> MultiServerMCPClient:
    """Create a MultiServerMCPClient that spawns figma-developer-mcp via stdio."""
    figma_api_key = os.environ.get("FIGMA_API_KEY", "")
    return MultiServerMCPClient(
        {
            "figma": {
                "transport": "stdio",
                "command": "npx",
                "args": [
                    "-y",
                    "figma-developer-mcp",
                    f"--figma-api-key={figma_api_key}",
                    "--stdio",
                ],
            }
        }
    )


async def _get_tool(tool_name: str) -> Any | None:
    """Spawn the figma-developer-mcp process and return the named tool."""
    tools = await _make_client().get_tools()
    return next((t for t in tools if t.name == tool_name), None)


async def figma_get_file(
    file_key: str,
    node_id: str | None = None,
    depth: int | None = None,
) -> dict[str, Any]:
    """Get a Figma file's layout and styling information.

    Retrieves design information from a Figma file, including layout structure,
    colors, typography, and component details. The response is simplified by
    figma-developer-mcp to include only the most relevant information.

    Args:
        file_key: The Figma file key (from the URL, e.g. "ABC123" in
            figma.com/file/ABC123/file-name). Also accepts a full Figma URL.
        node_id: Optional node ID to scope the fetch to a specific frame or
            component (e.g. "1:23").
        depth: Optional depth limit for nested elements (reduces response size
            for large files).

    Returns:
        Dictionary with keys:
        - success: Whether the request succeeded
        - document: Simplified layout and styling data
        - error: Error message if the request failed
    """
    try:
        tool = await _get_tool("get_figma_data")
        if not tool:
            return {
                "error": "get_figma_data tool not available from figma-developer-mcp",
                "success": False,
            }

        args: dict[str, Any] = {"fileKey": file_key}
        if node_id:
            args["nodeId"] = node_id
        if depth is not None:
            args["depth"] = depth

        result = await tool.ainvoke(args)
        return {"success": True, "document": result}

    except Exception as e:
        return {
            "error": f"Failed to get Figma file: {_unwrap_exception(e)}",
            "success": False,
        }


async def figma_get_component(
    file_key: str,
    component_id: str,
) -> dict[str, Any]:
    """Get detailed information about a specific Figma component or frame.

    Uses the node ID to scope the Figma data fetch to a single component,
    returning its layout, styling, and variant information.

    Args:
        file_key: The Figma file key containing the component.
        component_id: The node ID of the component (e.g. "1:23").

    Returns:
        Dictionary with keys:
        - success: Whether the request succeeded
        - component: Component layout and styling data
        - error: Error message if the request failed
    """
    try:
        tool = await _get_tool("get_figma_data")
        if not tool:
            return {
                "error": "get_figma_data tool not available from figma-developer-mcp",
                "success": False,
            }

        result = await tool.ainvoke({"fileKey": file_key, "nodeId": component_id})
        return {"success": True, "component": result}

    except Exception as e:
        return {
            "error": f"Failed to get Figma component: {_unwrap_exception(e)}",
            "success": False,
        }


async def figma_export_image(
    file_key: str,
    node_id: str,
    local_path: str = "/tmp/figma_images",
    file_name: str = "export.png",
) -> dict[str, Any]:
    """Download a Figma node (frame, component, or icon) as an image file.

    Downloads the specified node as SVG or PNG into a local directory inside
    the sandbox. Use the file path returned to reference the asset in code.

    Args:
        file_key: The Figma file key.
        node_id: The node ID to export (e.g. "1:23").
        local_path: Sandbox directory to write the image into (default:
            /tmp/figma_images). The directory must be writable.
        file_name: Output file name including extension, e.g. "button.svg"
            or "hero.png" (default: export.png).

    Returns:
        Dictionary with keys:
        - success: Whether the export succeeded
        - result: Server response confirming the downloaded file path
        - error: Error message if the export failed
    """
    try:
        tool = await _get_tool("download_figma_images")
        if not tool:
            return {
                "error": "download_figma_images tool not available from figma-developer-mcp",
                "success": False,
            }

        result = await tool.ainvoke(
            {
                "fileKey": file_key,
                "nodes": [{"nodeId": node_id, "fileName": file_name}],
                "localPath": local_path,
            }
        )
        return {"success": True, "result": result}

    except Exception as e:
        return {
            "error": f"Failed to export image: {_unwrap_exception(e)}",
            "success": False,
        }


async def close_figma_client() -> None:
    """No-op: stdio transport uses ephemeral per-call processes; nothing to close."""
    pass

File: src/tools/figmaMcp.ts
/**
 * Figma MCP tools for accessing Figma design files.
 * Mirrors agent/tools/figma_mcp.py
 *
 * Uses the figma-developer-mcp npm package launched via stdio transport.
 * Set FIGMA_API_KEY to a Figma Personal Access Token to enable the tools.
 */

import type { RunnableConfig } from "@langchain/core/runnables";
import { tool } from "@langchain/core/tools";
import { z } from "zod";

/**
 * Spawn a figma-developer-mcp process and invoke a tool by name.
 * Uses child_process since MCP stdio transport is process-based.
 */
async function invokeFigmaTool(
  toolName: string,
  args: Record<string, unknown>,
): Promise<unknown> {
  const { MultiServerMCPClient } = await import("@langchain/mcp-adapters");
  const figmaApiKey = process.env.FIGMA_API_KEY ?? "";

  const client = new MultiServerMCPClient({
    figma: {
      transport: "stdio",
      command: "npx",
      args: [
        "-y",
        "figma-developer-mcp",
        `--figma-api-key=${figmaApiKey}`,
        "--stdio",
      ],
    },
  });

  try {
    const tools = await client.getTools();
    const target = tools.find((t: { name: string }) => t.name === toolName);
    if (!target) {
      throw new Error(
        `${toolName} tool not available from figma-developer-mcp`,
      );
    }
    return await target.invoke(args);
  } finally {
    // MCP client may have a close method; if so, call it
    if (
      typeof (client as unknown as Record<string, unknown>).close === "function"
    ) {
      await (client as unknown as { close: () => Promise<void> }).close();
    }
  }
}

function unwrapException(e: unknown): string {
  if (e instanceof Error) {
    // Handle AggregateError / ExceptionGroup
    if ("errors" in e && Array.isArray((e as { errors: Error[] }).errors)) {
      return (e as { errors: Error[] }).errors
        .map((sub) => sub.message)
        .join("; ");
    }
    return e.message;
  }
  return String(e);
}

// ─── figma_get_file ──────────────────────────────────────────────────────────

const getFileSchema = z.object({
  fileKey: z
    .string()
    .describe(
      'The Figma file key (from the URL, e.g. "ABC123" in figma.com/file/ABC123/...). Also accepts a full Figma URL.',
    ),
  nodeId: z
    .string()
    .optional()
    .describe(
      "Optional node ID to scope the fetch to a specific frame or component (e.g. '1:23').",
    ),
  depth: z
    .number()
    .int()
    .optional()
    .describe("Optional depth limit for nested elements."),
});

export const figmaGetFile = tool(
  async ({ fileKey, nodeId, depth }, _config: RunnableConfig) => {
    try {
      const args: Record<string, unknown> = { fileKey };
      if (nodeId) args.nodeId = nodeId;
      if (depth !== undefined) args.depth = depth;

      const document = await invokeFigmaTool("get_figma_data", args);
      return JSON.stringify({ success: true, document });
    } catch (e) {
      return JSON.stringify({
        success: false,
        error: `Failed to get Figma file: ${unwrapException(e)}`,
      });
    }
  },
  {
    name: "figma_get_file",
    description:
      "Get a Figma file's layout and styling information. " +
      "Retrieves design information including layout structure, colors, typography, and component details.",
    schema: getFileSchema,
  },
);

// ─── figma_get_component ─────────────────────────────────────────────────────

const getComponentSchema = z.object({
  fileKey: z.string().describe("The Figma file key containing the component."),
  componentId: z
    .string()
    .describe("The node ID of the component (e.g. '1:23')."),
});

export const figmaGetComponent = tool(
  async ({ fileKey, componentId }, _config: RunnableConfig) => {
    try {
      const result = await invokeFigmaTool("get_figma_data", {
        fileKey,
        nodeId: componentId,
      });
      return JSON.stringify({ success: true, component: result });
    } catch (e) {
      return JSON.stringify({
        success: false,
        error: `Failed to get Figma component: ${unwrapException(e)}`,
      });
    }
  },
  {
    name: "figma_get_component",
    description:
      "Get detailed information about a specific Figma component or frame, " +
      "including layout, styling, and variant information.",
    schema: getComponentSchema,
  },
);

// ─── figma_export_image ──────────────────────────────────────────────────────

const exportImageSchema = z.object({
  fileKey: z.string().describe("The Figma file key."),
  nodeId: z.string().describe("The node ID to export (e.g. '1:23')."),
  localPath: z
    .string()
    .default("/tmp/figma_images")
    .describe("Sandbox directory to write the image into."),
  fileName: z
    .string()
    .default("export.png")
    .describe('Output file name including extension, e.g. "button.svg".'),
});

export const figmaExportImage = tool(
  async ({ fileKey, nodeId, localPath, fileName }, _config: RunnableConfig) => {
    try {
      const result = await invokeFigmaTool("download_figma_images", {
        fileKey,
        nodes: [{ nodeId, fileName }],
        localPath,
      });
      return JSON.stringify({ success: true, result });
    } catch (e) {
      return JSON.stringify({
        success: false,
        error: `Failed to export image: ${unwrapException(e)}`,
      });
    }
  },
  {
    name: "figma_export_image",
    description:
      "Download a Figma node (frame, component, or icon) as an image file. " +
      "Downloads the specified node as SVG or PNG into a local directory.",
    schema: exportImageSchema,
  },
);
